ambry-cloud/src/main/java/com.github.ambry.cloud/CloudBlobStore.java

```diff
@@ -207,8 +207,14 @@ public void updateTtl(MessageWriteSet messageSetToUpdate) throws StoreException
     // Note: we skipped uploading the blob on PUT record if the TTL was below threshold.
     try {
       for (MessageInfo msgInfo : messageSetToUpdate.getMessageSetInfo()) {
-        BlobId blobId = (BlobId) msgInfo.getStoreKey();
-        cloudDestination.updateBlobExpiration(blobId, msgInfo.getExpirationTimeInMs());
+        // MessageInfo.expirationTimeInMs is not reliable if ttlUpdate is set. See {@code PersistentIndex#findKey()}
+        // and {@code PersistentIndex#markAsPermanent()}.
+        if (msgInfo.isTtlUpdated()) {
+          BlobId blobId = (BlobId) msgInfo.getStoreKey();
+          cloudDestination.updateBlobExpiration(blobId, Utils.Infinite_Time);
+        } else {
+          logger.error("updateTtl() is called but msgInfo.isTtlUpdated is not set. msgInfo: {}", msgInfo);
```

Comment:
Wonder if this should be IllegalStateException.  If that is too severe, then add a VcrMetric error counter to track number of this or similar errors that are encountered.

Reply:
metric added.